### PR TITLE
Convert address utils from force text/bytes

### PR DESCRIPTION
**Pull request by njgheorghita**
@pipermerriam // @carver Thanks for the help on the other pr - broke it up to keep things simple but will use the old convo for reference.

### What was wrong?
Address utils used `force_bytes` / `force_text`

### How was it fixed?
- deprecated `force_bytes` / `force_bytes` and dependent functions
- converted address utils to use `to_bytes` / `to_text` from `conversions.py`
- dropped support for 32byte addresses

#### Cute Animal Picture
image

### CODE REVIEW

File: eth_utils/address.py [hunk 1]
```diff
@@ -145,38 +135,42 @@ def is_normalized_address(value):
...
+    returns it's canonical form (20 byte long string).
     """
-    return decode_hex(to_normalized_address(address))
+    text_address = to_text(primitive, hexstr, text)
+    return decode_hex(to_normalized_address(text_address))
 
 
-def is_canonical_address(value):
-    if not is_address(value):
+def is_canonical_address(address):
```
Comment (njgheorghita) [inline eth_utils/address.py]: I'm not sure if this behaves well with the new `is_address()` ? I can't seem to get the test case from the readme to pass (address=`b'\xd3\xcd\xa9\x13\xde\xb6\xf6yg\xb9\x9dg\xac\xdf\xa1q,)6\x01xd'`)

**Comment by pipermerriam** [comment]
@carver any chance you can spare cycles to code review this?

File: eth_utils/address.py [hunk 2]
```diff
@@ -35,25 +33,23 @@ def is_hex_address(value):
...
 def is_32byte_address(value):
     """
-    Checks if the given string is an address in hexidecimal encoded form padded to 32 bytes.
+    Checks if a given string is an address in hexidecimal encoded form padded to 32 bytes.
     """
-    if not is_string(value):
+    if not is_text(value):
         return False
 
     if len(value) == 32:
```
Comment (pipermerriam) [inline eth_utils/address.py]: I'm realizing this line is in contradiction to what the docstring for this function says.  @carver got any thoughts on this?  Should we support binary encoded addresses?  Should we split this function into two?

- `is_32byte_binary_address`
- `is_32byte_hex_address`

Not sure what the right thing to do here is, and honestly, I think **most** of the use cases for this function are theoretical.

File: eth_utils/address.py [hunk 3]
```diff
@@ -145,38 +135,42 @@ def is_normalized_address(value):
...
     if not is_address(left) or not is_address(right):
         raise ValueError("Both values must be valid addresses")
     else:
         return to_normalized_address(left) == to_normalized_address(right)
 
 
-@coerce_args_to_text
-@coerce_return_to_text
-def to_checksum_address(address):
+def to_checksum_address(primitive=None, hexstr=None, text=None):
```
Comment (pipermerriam) [inline eth_utils/address.py]: I don't think we can use the `primative/hexstr/text` signature here.  I actually think the original function *should* still work.  Is there something that breaks if this function is left unchanged?

File: eth_utils/address.py (hunk 1 above)
Comment (pipermerriam) [reply eth_utils/address.py]: This is due to the change to the signature of `is_checksum_address` and `to_checksum_address`.  I think if you revert the changes to those function signatures you'll be on the right track to fixing them.

File: eth_utils/address.py [hunk 4]
```diff
@@ -188,18 +182,21 @@ def to_checksum_address(address):
...
+    value = to_text(primitive, hexstr, text)
     if not is_hex_address(value):
         return False
-    return value == to_checksum_address(value)
+    return value == to_checksum_address(text=value)
 
 
-@coerce_args_to_text
-def is_checksum_formatted_address(value):
+def is_checksum_formatted_address(primitive=None, hexstr=None, text=None):
```
Comment (pipermerriam) [inline eth_utils/address.py]: I think this still needs to maintain a single `value` parameter as the only argument.

File: eth_utils/address.py [hunk 5]
```diff
@@ -188,18 +182,21 @@ def to_checksum_address(address):
...
-    return checksum_address
+    if is_bytes(checksum_address):
+        return to_text(primitive=checksum_address)
+    else:
+        return to_text(text=checksum_address)
 
 
-@coerce_args_to_text
-def is_checksum_address(value):
+def is_checksum_address(primitive=None, hexstr=None, text=None):
```
Comment (pipermerriam) [inline eth_utils/address.py]: I think this still needs to maintain a single `value` parameter as the only argument.

File: eth_utils/address.py (hunk 1 above)
Comment (njgheorghita) [reply eth_utils/address.py]: I guess i'm a little confused on what exactly a canonical representation is. Isn't it a 20byte long string of bytes? @carver suggested in the last pr the below implementation and "We don't need to test the canonical-ness of bytes, it's clearly not. "
```python
def is_canonical_address(address):
    if not is_string(address) or not is_hex(address):
        return False
    elif not is_address(hexstr=address):
        return False
    else:
        return address == to_canonical_address(hexstr=address)
```
The test case from the readme passes now but breaks if I add `if not is_string(address) or not is_hex(address):`

Comment (pipermerriam) [reply eth_utils/address.py]: Canonical address definition is a byte-string of length 20.  Feel free to modify this function to be *loosely* the following.

```python
def is_canonical_address(value):
    return is_bytes(value) and len(value) == 20
```

Comment (carver) [reply eth_utils/address.py]: Ah, interesting, yes I misinterpreted `canonical`. The "typical" representation is checksum, so I was basically reading it as checking if the address is the checksum. I think the same things I was saying before still apply, you can just reject it if it's not `bytes`, rather than convert `str` to `bytes` and check if that `is_address`, etc.

File: eth_utils/address.py [hunk 6]
```diff
@@ -188,18 +177,25 @@ def to_checksum_address(address):
         )
         for i in range(2, 42)
     ))
-    return checksum_address
+    if is_bytes(checksum_address):
```
Comment (carver) [inline eth_utils/address.py]: `add_0x_prefix` should never be returning `bytes`. Why can't we just return `checksum_address` here?

File: eth_utils/string.py [hunk 7]
```diff
@@ -1,15 +1,33 @@
...
+            "some byte values in the 0-255 range which makes naive coersion between "
+            "bytes and text representations impossible without explicitly "
+            "declared encodings.".format(fn.__name__)
+        ))
+        warnings.resetwarnings()
+        return fn(*args, **kwargs)
+    return inner
+
+
+@_deprecated
```
Comment (carver) [inline eth_utils/string.py]: This deprecation warning should be submitted as a separate PR against the v0 branch. They can just be deleted in v1 (or if all usages aren't removed in this PR, then left in only until the cleanup is done).

File: tests/address-utils/test_address_utils.py [hunk 8]
```diff
@@ -77,16 +100,19 @@ def test_is_checksum_address(value, expected):
...
         ('0xD1220A0cf47c7B9Be7A2E6BA89F429762e7b9aDb', True),
+        ('0x52908400098527886E0F7030069857D2E4169EE7', False),
+        ('0x8617E340B3D01FA5F11F306F4090FD50E238070D', False),
+        ('0xde709f2102306220921060314715629080e2fb77', False),
+        ('0x27b1fdb04752bbc536007a920d24acb045561c26', False),
         ('0XD1220A0CF47C7B9BE7A2E6BA89F429762E7B9ADB', False),
-        ('0xd1220a0cf47c7b9be7a2e6ba89f429762e7b9adb', False)
+        ('0xd1220a0cf47c7b9be7a2e6ba89f429762e7b9adb', False),
+        (b'0XD1220A0CF47C7B9BE7A2E6BA89F429762E7B9ADB', False),
+        (b'0xd1220a0cf47c7b9be7a2e6ba89f429762e7b9adb', False),
```
Comment (carver) [inline tests/address-utils/test_address_utils.py]: Another nice case would be a checksum address in `bytes` that would otherwise pass if it were decoded to `str`, like `b'0x5aAeb6053F3E94C9b9A09f33669435E7Ef1BeAed'`.

File: tests/address-utils/test_address_utils.py [hunk 9]
```diff
@@ -96,14 +122,18 @@ def test_is_checksum_formatted_address(value, expected):
 @pytest.mark.parametrize(
     "value,expected",
     [
-        # weird values
-        (lambda: None, False),
-        ("function", False),
-        ({}, False),
```
Comment (carver) [inline tests/address-utils/test_address_utils.py]: Why delete these? It seems an unlikely scenario, but they're not slow tests or anything.

File: tests/address-utils/test_address_utils.py [hunk 10]
```diff
@@ -146,11 +160,11 @@ def test_is_normalized_address(value, expected):
             '0xc305c901078781c232a2a521c2af7980f8385ee9',
         ),
         (
-            b'0x000000000000000000000000c305c901078781c232a2a521c2af7980f8385ee9',
-            '0xc305c901078781c232a2a521c2af7980f8385ee9',
+            'C6D9D2CD449A754C494264E1809C50E34D64562B',
+            '0xc6d9d2cd449a754c494264e1809c50e34d64562b',
```
Comment (carver) [inline tests/address-utils/test_address_utils.py]: duplicate case from two above?

File: tests/address-utils/test_address_utils.py [hunk 11]
```diff
@@ -238,28 +249,22 @@ def test_is_same_address(address1, address2, expected):
...
 
 @pytest.mark.parametrize(
     'value,expected',
     (
-        (b'\xd3\xcd\xa9\x13\xde\xb6\xf6yg\xb9\x9dg\xac\xdf\xa1q,)6\x01', True),
-        ('\xd3\xcd\xa9\x13\xde\xb6\xf6yg\xb9\x9dg\xac\xdf\xa1q,)6\x01', False),
-        ('0xd3cda913deb6f67967b99d67acdfa1712c293601', False),
+	('\xd3\xcd\xa9\x13\xde\xb6\xf6yg\xb9\x9dg\xac\xdf\xa1q,)6\x01', False),
+	(b'\xd3\xcd\xa9\x13\xde\xb6\xf6yg\xb9\x9dg\xac\xdf\xa1q,)6\x01', True),
+	(b'0xd3cda913deb6f67967b99d67acdfa1712c293601', False),
```
Comment (carver) [inline tests/address-utils/test_address_utils.py]: Should probably leave in the `str` version of this case.

File: eth_utils/address.py (hunk 1 above)
Comment (njgheorghita) [reply eth_utils/address.py]: I'm thinking something like . . . ?
```python
if not is_bytes(address) or len(address) != 20:
    return False
return address == to_canonical_address(address)

File: eth_utils/address.py (hunk 2 above)
Comment (njgheorghita) [reply eth_utils/address.py]: Thanks for the help @carver and @pipermerriam ! Waiting on confirmation whether/not to drop 32byte support from Piper, and I'll push the updated version.

File: eth_utils/string.py (hunk 7 above)
Comment (njgheorghita) [reply eth_utils/string.py]: Made the pr to v0 branch. I plan to submit the cleanup prs 1 by 1 - and will delete them on the last one.

File: eth_utils/address.py (hunk 2 above)
Comment (njgheorghita) [reply eth_utils/address.py]: @pipermerriam thoughts on dropping support for 32-byte addresses?

**Comment by njgheorghita** [comment]
Thanks for the help @pipermerriam & @carver, appreciate it. Took care of your comments - let me know if there's anything else that needs adjusting

**Comment by njgheorghita** [comment]
@pipermerriam ping!

**Comment by pipermerriam** [comment]
@njgheorghita pong

**Comment by njgheorghita** [comment]
@pipermerriam 😜 gracias!
